plugin/pkg/managedseed/validator/admission.go

```diff
@@ -0,0 +1,373 @@
+	PluginName = "ManagedSeedValidator"
+)
+
+// Register registers a plugin.
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
+		return New()
+	})
+}
+
+// ValidateManagedSeed contains listers and and admission handler.
+type ValidateManagedSeed struct {
+	*admission.Handler
+	coreClient coreclientset.Interface
+	kubeClient kubernetes.Interface
+	readyFunc  admission.ReadyFunc
+}
+
+var (
+	_ = admissioninitializer.WantsInternalCoreClientset(&ValidateManagedSeed{})
+	_ = admissioninitializer.WantsKubeClientset(&ValidateManagedSeed{})
+
+	readyFuncs = []admission.ReadyFunc{}
+)
+
+// New creates a new ValidateManagedSeed admission plugin.
+func New() (*ValidateManagedSeed, error) {
+	return &ValidateManagedSeed{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+	}, nil
+}
+
+// AssignReadyFunc assigns the ready function to the admission handler.
+func (v *ValidateManagedSeed) AssignReadyFunc(f admission.ReadyFunc) {
+	v.readyFunc = f
+	v.SetReadyFunc(f)
+}
+
+// SetInternalCoreClientset sets the garden core clientset.
+func (v *ValidateManagedSeed) SetInternalCoreClientset(c coreclientset.Interface) {
+	v.coreClient = c
+}
+
+// SetKubeClientset gets the clientset from the Kubernetes client.
+func (v *ValidateManagedSeed) SetKubeClientset(c kubernetes.Interface) {
+	v.kubeClient = c
+}
+
+// ValidateInitialization checks whether the plugin was correctly initialized.
+func (v *ValidateManagedSeed) ValidateInitialization() error {
+	if v.coreClient == nil {
+		return errors.New("missing garden core client")
+	}
+	if v.kubeClient == nil {
+		return errors.New("missing kubernetes client")
+	}
+	return nil
+}
+
+var _ admission.MutationInterface = &ValidateManagedSeed{}
+
+// Admit validates and if appropriate mutates the given managed seed against the shoot that it references.
+func (v *ValidateManagedSeed) Admit(ctx context.Context, a admission.Attributes, o admission.ObjectInterfaces) error {
+	// Wait until the caches have been synced
+	if v.readyFunc == nil {
+		v.AssignReadyFunc(func() bool {
+			for _, readyFunc := range readyFuncs {
+				if !readyFunc() {
+					return false
+				}
+			}
+			return true
+		})
+	}
+	if !v.WaitForReady() {
+		return admission.NewForbidden(a, errors.New("not yet ready to handle request"))
+	}
+
+	// Ignore all kinds other than Shoot
+	if a.GetKind().GroupKind() != seedmanagement.Kind("ManagedSeed") {
+		return nil
+	}
+
+	// Ignore updates to status or other subresources
+	if a.GetSubresource() != "" {
+		return nil
+	}
+
+	// Convert object to ManagedSeed
+	managedSeed, ok := a.GetObject().(*seedmanagement.ManagedSeed)
+	if !ok {
+		return apierrors.NewBadRequest("could not convert object to ManagedSeed")
+	}
+
+	// Get shoot
+	shoot, err := v.coreClient.Core().Shoots(managedSeed.Namespace).Get(ctx, managedSeed.Spec.Shoot.Name, clientkubernetes.DefaultGetOptions())
```

Comment:
OK, sure. I did like this at first but then decided to make it "simpler", not thinking about the performance impact. A lister would read many resources at once and cache them, am I getting this correctly?

Reply:
From your other comment where you pointed out that a lister is wrong because we might miss updates - it's not very clear to me in when is it safe to use a lister, and when is it not?

Reply:
> A lister would read many resources at once and cache them, am I getting this correctly?

Yes, the lister will build up a cache for all shoots. Though, all admission plugins share the same cache and we already have some plugins reading from the cache.
So, it's rather the opposite, that the reading client will have a negative performance impact (because of net I/O) and the lister will be more performant and won't degrade any performance.

We should first consult the lister. If the shoot is not present in the cache yet, we should do a direct lookup though, because `ManagedSeed` and `Shoot` might have been created shortly after one another (similar to how we check the secretRef in SecretBindings).

> From your other comment where you pointed out that a lister is wrong because we might miss updates - it's not very clear to me in when is it safe to use a lister, and when is it not?

Ok. Generally speaking, it's more desirable to read from the cache instead of calling the API.
But, we should be careful with reading from the cache for operations, were it would be critical if we have missed some updates, for example when checking a resource in some deletion protection mechanism.
This is probably a good rule of thumb.

Reply:
OK, understood. Does it mean we should do this "double search" (first via the lister, then via the client) each and every time, or only in specific circumstances?

Reply:
Only in specific circumstances. For example, if we could potentially deny a valid request, e.g. the case where SecretBinding and Secret are created simultaneously and the cache might not have synced the Secret fast enough. Same applies for the ManagedSeed and Shoot creation.

Reply:
OK, thanks, I will adapt the code according to your recommendations.